crates/functional/src/lib.rs

```diff
@@ -0,0 +1,247 @@
+                            ));
+                            true
+                        },
+                    );
+                }),
+            );
+        },
+        move || UseReducerState {
+            current_state: Rc::new(init(initial_state)),
+        },
+    );
+    return ret;
+}
+
+pub fn use_state<T, F>(initial_state_fn: F) -> (Rc<T>, Box<impl Fn(T)>)
+where
+    F: FnOnce() -> T,
+    T: 'static,
+{
+    struct UseStateState<T2> {
+        current: Rc<T2>,
+    }
+    return use_hook(
+        |prev: &mut UseStateState<T>, hook_update| {
+            let current = prev.current.clone();
+            return (
+                current,
+                Box::new(move |o: T| {
+                    hook_update(|state: &mut UseStateState<T>| {
+                        state.current = Rc::new(o);
+                        true
+                    });
+                }),
+            );
+        },
+        move || UseStateState {
+            current: Rc::new(initial_state_fn()),
+        },
+    );
+}
+
+pub fn use_hook<InternalHookState, HookRunner, R, InitialStateProvider, PretriggerChange: 'static>(
+    hook_runner: HookRunner,
+    initial_state_producer: InitialStateProvider,
+) -> R
+where
+    HookRunner: FnOnce(&mut InternalHookState, Box<dyn Fn(PretriggerChange)>) -> R,
+    InternalHookState: 'static,
+    InitialStateProvider: FnOnce() -> InternalHookState,
+    PretriggerChange: FnOnce(&mut InternalHookState) -> bool,
+{
+    // Extract current hook
+    let (mut component_hook_state, hook) = CURRENT_HOOK.with(|hook_state_holder| {
+        let component_hook_state = hook_state_holder.clone();
+        let hook_state_holder = hook_state_holder.try_borrow_mut();
+        let mut hook_state_holder = hook_state_holder.expect("Nested hooks not supported");
+        let mut hook_state = hook_state_holder
+            .as_mut()
+            .expect("No current hook. Hooks can only be called inside functional components");
+
+        // Determine which hook position we're at and increment for the next hook
+        let hook_pos = hook_state.counter;
+        hook_state.counter += 1;
+
+        // Initialize hook if this is the first call
+        if hook_pos >= hook_state.hooks.len() {
+            let initial_state = Rc::new(RefCell::new(initial_state_producer()));
+            hook_state.hooks.push(initial_state);
+        }
+
+        let hook = hook_state.hooks[hook_pos].clone();
+
+        return (component_hook_state, hook);
+    });
+
+    let trigger = {
+        let hook = hook.clone();
+        Box::new(move |pretrigger_change: PretriggerChange| {
+            let mut hook = hook.clone();
+            (component_hook_state
+                .try_borrow_mut()
+                .expect("Hook is being modified elsewhere. Note that hooks cannot be nested.")
+                .deref_mut()
+                .as_mut()
+                .unwrap()
+                .process_message)(Box::new(move || {
```

Comment:
Any reason why you aren't cloning `process_message` when extracting the current hook and using it here?

Reply:
Lapse in concentration ;)

Reply:
Ok, I'll let this one slide! 😝